Add Full-duplex serial driver for ARM boards

KarlK90 opened the pull request:
Hi there first time contributor here :wave:

## Description

This pull request adds a new Full-duplex serial driver for chibios based ARM split keyboards. I started this because "my yet to be" released keyboard has the hardware capabilities to connect the rx and tx pins of each half to the corresponding tx and rx pins of the other half. While reading the `serial_usart.c` code I also noticed that the receive and send buffers have to be cleared for every transaction, which is a limitation of single-wire Half-duplex as the source code stated. This driver tries to improve on the existing Half-duplex driver with the downside that it requires one more conductor:

* No need to flush the send and receive buffers after every transaction
* No need for external pull-up resistors as the tx pin is in push-pull mode
* UART driver in blocking mode to use DMA transfers (not fully asynchronous but hey :shrug:)

The driver is implemented in `drivers/chibios/serial_usart_duplex.c`.

## Refactor 

Most of the changes are renames of the existing `soft_serial` naming scheme to just `serial`. My reasoning is that `soft_serial` is a legacy from the time where qmk only had the bitbang serial implementation, which changed with the introduction of the usart backed implementation on arm. Therefore I decided to rename the occurrences of soft_serial to just serial namely: 

```
SELECT_SOFT_SERIAL_SPEED -> SELECT_SERIAL_SPEED
SOFT_SERIAL_PIN -> SERIAL_PIN_TX
soft_serial_initiator_init -> serial_initiator_init
soft_serial_target_init -> serial_target_init
soft_serial_transaction -> serial_transaction
soft_serial_get_and_clean_status -> serial_get_and_clean_status
handle_soft_serial_slave -> handle_serial_slave
```

The renames where done as separate commits.  I'am pretty unsure if this changed is to intrusive but I hope this resonates well with the qmk community nevertheless :slightly_smiling_face: 

### Measurements
Setup:
* 8x5 split keyboard
* 33xWS2812 RGB LEDs on each half (PWM Driver)
* GD32F103CB with 72MHz SYSCLK and 72MHz PCLK2
* Ofast optimizations and LTO enabled

Results:
* USART is stable at 2.250.000 Baud/s, 4.500.000 Bauds/s has too many timeouts
* One complete transaction between initiator and target takes 67us
* Matrix scan frequency is about 3800 idle with RGB Lighting on
* Matrix scan frequency is about 4000 idle with RGB Lighting off

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* None

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented:
You need to rebase your branch locally against the qmk `develop` branch.  
If you need any assistance with just, just ask.

KarlK90 commented:
Rebased against develop and added a small cosmetic refactoring commit.

noroadsleft commented:
This needs to be rebased again, and have the merge conflicts resolved.

KarlK90 commented:
Rebased onto latest develop branch

zvecr commented on drivers/chibios/serial_usart_duplex.c:
```diff
@@ -0,0 +1,245 @@
+    palSetLineMode(SERIAL_USART_RX_PIN, PAL_MODE_INPUT);
+#else
+    palSetLineMode(SERIAL_USART_TX_PIN, PAL_MODE_ALTERNATE(SERIAL_USART_TX_PAL_MODE) | PAL_STM32_OTYPE_PUSHPULL | PAL_STM32_OSPEED_HIGHEST);
+    palSetLineMode(SERIAL_USART_RX_PIN, PAL_STM32_MODE_INPUT);
```
This would also need to handle AF mode as per TX pin

KarlK90 replied:
Could you elaborate on this? To my understanding the TX and RX pins are already correctly configured.

KarlK90 replied:
Introduced SERIAL_USART_RX_PAL_MODE for recent mcus like the f303 only had experience with the f103 and see the point now.

tzarc commented:
> We previously tested full duplex while implementing half duplex, and found that it was not any faster.
For reference, [earlier attempted implementation using TX/RX swap
](https://github.com/tzarc/qmk_build/blob/master/alternates/serial_usart_fullduplex.c).
There's also an earlier (broken) UART driver implementation in the same directory, which got shelved because of the following issue...

It's my understanding that the ChibiOS serial driver wrapped some internal buffer management and the RX interrupt to ensure all data is received -- which was why the Serial driver was used in preference of the UART driver. Conceivably, if there's no "active" call to `uartReceiveTimeout` in progress (hypothetically say a thread context switch to RGB updates before the next call), any data received would be dropped?

KarlK90 commented:
Rebased and pushed to latest develop branch. Sorry for letting this go stall for such a long time.

KarlK90 commented:
> Thanks for the contribution, its an area we have previously discussed a fair bit. So a few things,
> 
>     1. The Rename refactor should be its own ticket, to which we have previously talked about picking a different name than "serial"
>        
>        * The decision on previous discussions was to avoid renaming in the short term to avoid further confusion.
> 
>     2. We previously tested full duplex while implementing half duplex, and found that it was not any faster.
>        
>        * The main advantage I can see is removing the pullup requirement, though the current suggestion is to run half duplex on the same pins as i2c so that you might be able to select that method in the future.
>        * The downside is the amount code duplication which is introduced. We also tinkered with rx<->tx swap which works on a limited amount of chips, as another method of i2c interop.
> 
>     3. Current terminology for the project is still master/slave for the time being, while we decide on #9376. Introducing changes to areas such as the docs should be avoided for now to reduce fragmentation on the issue.

1. Reverted the renaming commits, but used `SERIAL_USART_TX_PIN` and `SERIAL_USART_RX_PIN` instead of `SOFT_SERIAL_PIN` for the driver as it seemed more appropriate then introducing `SOFT_SERIAL_PIN_RX`.
2. I have to admit that there was no benchmarking involved while claiming those improvements :see_no_evil:. Pin swapping is handled using the define `SERIAL_USART_PIN_SWAP` for this driver. 
3. Reverted the changes and used master/slave terminology. But I really would like to change that in the future.

KarlK90 commented:
> > We previously tested full duplex while implementing half duplex, and found that it was not any faster.
> 
> For reference, [earlier attempted implementation using TX/RX swap ](https://github.com/tzarc/qmk_build/blob/master/alternates/serial_usart_fullduplex.c)
> There's also an earlier (broken) UART driver implementation in the same directory, which got shelved because of the following issue...
> 
> It's my understanding that the ChibiOS serial driver wrapped some internal buffer management and the RX interrupt to ensure all data is received -- which was why the Serial driver was used in preference of the UART driver. Conceivably, if there's no "active" call to `uartReceiveTimeout` in progress (hypothetically say a thread context switch to RGB updates before the next call), any data received would be dropped?

I think that situation is unlikely to occur because there is no other thread than the main and idle thread right now and the receiving thread has the highest priority of them all. So in theory there should be no context switch back to any other thread until `uartReceiveTimeout` was called again and the receiving thread put to sleep? Another thought is that everything is synced many thousand times per second, so even if one transfer failed there are plenty of backup transfers. What do you think?

tzarc commented:
> I think that situation is unlikely to occur because there is no other thread than the main and idle thread right now

Untrue for USB-OTG MCUs -- USB runs in a thread inside ChibiOS.

KarlK90 commented:
Introduced `SERIAL_USART_TX_PAL_MODE` for recent MCUs. I will be able to verify the changes in the coming days, but If anyone has a ready f072 or f303 keyboard with duplex usart comms to test it, please do so :-).

tzarc commented:
Sorry about that, GitHub decided to delete the `develop` branch from upstream when we merged it, despite being told otherwise. Reopened.

KarlK90 commented:
@tzarc @zvecr The latest change takes care of the missed handshakes by using the rxchar callback function. Please let me know if there is anything more that prevents this driver from merging. :slightly_smiling_face:

drashna commented on docs/serial_driver.md:
````diff
@@ -68,3 +70,167 @@ You must also enable the ChibiOS `SERIAL` feature:
+
+```c
+// Remap USART1 from pins PA9/PA10 to PB6/PB7
+void keyboard_pre_init_user(void) {
````
Would rather this be a setting, via a define, than this.

KarlK90 replied:
Add defines for USART remapping. This affects half-duplex as well as full-duplex.

drashna commented on docs/serial_driver.md:
```diff
@@ -68,3 +70,167 @@ You must also enable the ChibiOS `SERIAL` feature:
+
+Do note that the configuration required is for the `UART` peripheral, not the `SERIAL` peripheral.
+
+#### Pins for USART Peripherals with Alternate Functions for selected STM32 MCUs
```
Would be nice if all of this was a single table, though I know what may be complex to do.

KarlK90 replied:
Merged the tables into a single one for each mcu. Having one big table for all mcus is not practical imho.

KarlK90 commented:
Rebased onto develop and addressed @drashna requests.

drashna commented on drivers/chibios/serial_usart_duplex.c:
```diff
@@ -0,0 +1,306 @@
+#    define SELECT_SERIAL_SPEED 1
+#endif
+
+#if defined(SERIAL_USART_SPEED)
```
A lot of this looks to be identical, or very close to the half duplex driver.  It may be worth pulling out the shared config and putting them in an h file to be shared

KarlK90 replied:
I have moved all shared configurations into serial_usart.h and added QMK copyright headers.

KarlK90 commented:
@zvecr I have added copyright headers mentioning QMK as copyright holder to `serial_usart.c` is that fine with you?

tzarc commented on docs/serial_driver.md:
```diff
@@ -68,3 +71,140 @@ You must also enable the ChibiOS `SERIAL` feature:
+
+#### Pin configuration
+
+`SERIAL_USART_TX_PIN` is the USART `TX` pin, `SERIAL_USART_RX_PIN` is the USART `RX` pin. No external pull-up resistors are needed as the `TX` pin operates in push-pull mode. To use this driver the usart peripherals `TX` and `RX` pins must be configured with the correct Alternate-functions. If you are using an STM32F103 everything is already setup, for other MCUs which are using a more modern GPI ...
```
Pretty sure we target Proton-C/F303 by default. Will need to confirm defaults elsewhere.
```suggestion
`SERIAL_USART_TX_PIN` is the USART `TX` pin, `SERIAL_USART_RX_PIN` is the USART `RX` pin. No external pull-up resistors are needed as the `TX` pin operates in push-pull mode. To use this driver the usart peripherals `TX` and `RX` pins must be configured with the correct Alternate-functions. If you are using a Proton-C, everything is already setup, for other MCUs which are using a more modern GPIO configuration you have to specify these by setting `SERIAL_USART_TX_PAL_MODE` and `SERIAL_USART_RX_PAL_MODE`. Refeer to the corresponding datasheets of your MCU or find those Settings in the table below.
```

KarlK90 replied:
Added Proton-C to the sentence and adjusted the following, not mentioning the stm32f103 would have skewed the meaning as If the f303 wouldn't have a modern gpio configuration.

tzarc commented on docs/serial_driver.md:
```diff
@@ -68,3 +71,140 @@ You must also enable the ChibiOS `SERIAL` feature:
+`SERIAL_USART_TX_PIN` is the USART `TX` pin, `SERIAL_USART_RX_PIN` is the USART `RX` pin. No external pull-up resistors are needed as the `TX` pin operates in push-pull mode. To use this driver the usart peripherals `TX` and `RX` pins must be configured with the correct Alternate-functions. If you are using an STM32F103 everything is already setup, for other MCUs which are using a more modern GPI ...
+
+#### Connecting the halves and Pin Swap
+Please note that `TX` of the master halve has to be connected with the `RX` pin of the slave halve and `RX` of the master halve has to be connected with the `TX` pin of the slave halve! Usually this pin swap has to be done outside of the MCU e.g. with cables or on the pcb. Some MCUs like the STM32F303 allow this pin swap directly inside the MCU, this feature can be enabled using `#define SERIAL_U ...
```
```suggestion
Please note that `TX` of the master half has to be connected with the `RX` pin of the slave half and `RX` of the master half has to be connected with the `TX` pin of the slave half! Usually this pin swap has to be done outside of the MCU e.g. with cables or on the pcb. Some MCUs like the STM32F303 allow this pin swap directly inside the MCU, this feature can be enabled using `#define SERIAL_USART_PIN_SWAP` in your config.h.
```

tzarc commented on docs/serial_driver.md:
```diff
@@ -68,3 +71,140 @@ You must also enable the ChibiOS `SERIAL` feature:
+//#define USART1_REMAP             // Remap USART TX and RX pins on STM32F103 mcus, see table below.
+//#define SERIAL_USART_PIN_SWAP    // Swap TX and RX pins if keyboard is master halve.
+                                   // Check if this feature is necessary with your keyboard design and available on the mcu.
+#define SELECT_SOFT_SERIAL_SPEED 1 // or 0, 2, 3, 4, 5
+                                   //  0: about 460800 baud
+                                   //  1: about 230400 baud (default)
+                                   //  2: about 115200 baud
+                                   //  3: about 57600 baud
+                                   //  4: about 38400 baud
+                                   //  5: about 19200 baud
```
```suggestion
#define SELECT_SOFT_SERIAL_SPEED 1 // or 0, 2, 3, 4, 5
                                   //  0: 460800 baud
                                   //  1: 230400 baud (default)
                                   //  2: 115200 baud
                                   //  3: 57600 baud
                                   //  4: 38400 baud
                                   //  5: 19200 baud
```

tzarc commented on docs/serial_driver.md:
```diff
@@ -68,3 +71,140 @@ You must also enable the ChibiOS `SERIAL` feature:
+| PD8        | TX       | AFPP | USART3_FULLREMAP    |
+| PD9        | RX       | IN   | USART3_FULLREMAP    |
+
+##### STM32F303 [Datasheet](https://www.st.com/resource/en/datasheet/stm32f303cc.pdf)
```
Move this to be first, given the Proton-C focus.

tzarc commented on docs/serial_driver.md:
````diff
@@ -68,3 +71,140 @@ You must also enable the ChibiOS `SERIAL` feature:
+```
+
+You must also enable the ChibiOS `UART` with blocking api feature:
+* In your board's halconf.h: `#define HAL_USE_USAR TRUE` and `#define UART_USE_WAIT TRUE`
````
```suggestion
* In your board's halconf.h: `#define HAL_USE_UART TRUE` and `#define UART_USE_WAIT TRUE`
```

tzarc commented on drivers/chibios/serial_usart.h:
```diff
@@ -0,0 +1,78 @@
+#    define SERIAL_USART_CR3 0
+#endif
+
+#if defined(USART1_REMAP)
+#    define USART_REMAP (AFIO->MAPR |= AFIO_MAPR_USART1_REMAP);
+#elif defined(USART2_REMAP)
+#    define USART_REMAP (AFIO->MAPR |= AFIO_MAPR_USART2_REMAP);
+#elif defined(USART3_PARTIALREMAP)
+#    define USART_REMAP (AFIO->MAPR |= AFIO_MAPR_USART3_REMAP_PARTIALREMAP);
+#elif defined(USART3_FULLREMAP)
+#    define USART_REMAP (AFIO->MAPR |= AFIO_MAPR_USART3_REMAP_FULLREMAP);
+#endif
```
```suggestion
#if defined(USART1_REMAP)
#    define USART_REMAP do { (AFIO->MAPR |= AFIO_MAPR_USART1_REMAP); } while(0)
#elif defined(USART2_REMAP)
#    define USART_REMAP do { (AFIO->MAPR |= AFIO_MAPR_USART2_REMAP); } while(0)
#elif defined(USART3_PARTIALREMAP)
#    define USART_REMAP do { (AFIO->MAPR |= AFIO_MAPR_USART3_REMAP_PARTIALREMAP); } while(0)
#elif defined(USART3_FULLREMAP)
#    define USART_REMAP do { (AFIO->MAPR |= AFIO_MAPR_USART3_REMAP_FULLREMAP); } while(0)
#endif
```

tzarc commented on drivers/chibios/serial_usart.c:
```diff
@@ -123,6 +90,10 @@ __attribute__((weak)) void usart_init(void) {
 #endif
+
+#if defined(USART_REMAP)
+    USART_REMAP
```
```suggestion
    USART_REMAP;
```

tzarc commented on drivers/chibios/serial_usart_duplex.c:
```diff
@@ -0,0 +1,261 @@
+    usart_init();
+
+#if defined(USART_REMAP)
+    USART_REMAP
```
```suggestion
    USART_REMAP;
```

tzarc commented on drivers/chibios/serial_usart_duplex.c:
```diff
@@ -0,0 +1,261 @@
+#endif
+
+#if defined(USART_REMAP)
+    USART_REMAP
```
```suggestion
    USART_REMAP;
```

tzarc commented:
Confirmed working with a pair of Proton-C boards.

KarlK90 commented:
Applied the Code Review suggestions, thank you!

KarlK90 commented:
Should squash the commits once this pr has been accepted?

tzarc commented:
> 
> 
> Should squash the commits once this pr has been accepted?

GitHub will do so when we merge.

drashna reviewed (approved):
Compiles fine locally.

KarlK90 commented on docs/serial_driver.md:
```diff
@@ -68,3 +71,140 @@ You must also enable the ChibiOS `SERIAL` feature:
 Do note that the configuration required is for the `SERIAL` peripheral, not the `UART` peripheral.
+
+### USART Full-duplex
+Targeting STM32 boards where communication is offloaded to a USART hardware device. The advantage over bitbang is that this provides fast and accurate timings. USART Full-Duplex requires two conductors instead of one conductor unlike the Half-duplex driver, but it is more efficent as it uses DMA transfers, which can result in even faster transmission speeds.
```
No external pullup is needed. I could run with dual open drain pins, which requires two pull up resistors. That is not the default though.

KarlK90 replied:
Updated the wording.

drashna replied:
Awesome!

XScorpion2 commented:
Confirmed this works really well on the Mun
